### PR TITLE
Add Ohio Joint Filing Credit

### LINKED ISSUE #3057: Ohio Joint Filing Credit
Reference:

https://tax.ohio.gov/static/ohio_individual/individual/SupportingDocumentation/Pages/SoC-12_JFC.html

https://tax.ohio.gov/static/forms/ohio_individual/individual/2022/it1040-bundle.pdf

https://tax.ohio.gov/static/forms/ohio_individual/individual/2022/it1040-sd100-instruction-booklet.pdf

### PR DESCRIPTION
**Pull request by mylittlebecca**
Fix #3057

### CODE REVIEW

File: policyengine_us/variables/gov/states/oh/tax/income/credits/oh_joint_filing_credit.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        filing_status = tax_unit("filing_status", period)
+        credit = tax_unit("oh_income_tax_before_refundable_credits", period)
+        person = tax_unit.members
+        is_head = person("is_tax_unit_head", period)
+        is_spouse = person("is_tax_unit_spouse", period)
+        p = parameters(
+            period
+        ).gov.states.oh.tax.income.credits.joint_filing_credit
+        income = person("employment_income", period)
+        print(income)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/oh_joint_filing_credit.py:23]: delete prints

File: policyengine_us/variables/gov/states/oh/tax/income/credits/oh_joint_filing_credit.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from policyengine_us.model_api import *
+
+
+class oh_joint_filing_credit(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/oh_joint_filing_credit.py:4]: make a oh_joint_filing_credit_eligible.py with 
qualify_income & qualify_status

File: policyengine_us/tests/policy/baseline/gov/states/oh/tax/income/credits/oh_joint_filing_credit.yaml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        employment_income: 600
+      person2:
+        is_tax_unit_spouse: true
+        employment_income: 700
+    tax_units:
+      tax_unit:
+        members: [person1, person2]
+        filing_status: JOINT
+        oh_agi: 47_000
+        oh_income_tax_exempt: 10_000
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/oh/tax/income/credits/oh_joint_filing_credit.yaml:59]: need to add state code OH

File: policyengine_us/variables/gov/states/oh/tax/income/oh_income_tax_before_refundable_credits.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/oh_income_tax_before_refundable_credits.py:1]: no test

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+from policyengine_us.model_api import *
+
+
+class oh_joint_filing_credit_eligible(Variable):
+    value_type = float
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:5]:
```suggestion
    value_type = bool
```

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        p = parameters(
+            period
+        ).gov.states.oh.tax.income.credits.joint_filing_credit
+        income = person("employment_income", period)
+
+        head_income = tax_unit.sum(is_head * income)
+
+        spouse_income = tax_unit.sum(is_spouse * income)
+
+        qualify_income = (head_income >= p.income_base) & (
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:26]:
```suggestion
        income_eligible = (head_income >= p.income_base) & (
```

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            spouse_income >= p.income_base
+        )
+        qualify_status = filing_status == filing_status.possible_values.JOINT
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:29]:
```suggestion
        status_eligible = filing_status == filing_status.possible_values.JOINT
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_base.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+description: Ohio qualifies the joint filing credit for head and spouse each with Ohio adjusted gross income above this amount. 
+values:
+  2021-01-01: 500
+
+metadata:
+  period: year
+  unit: currency-USD
+  label: Ohio joint filing credit base qualifying income
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_base.yaml:8]:
```suggestion
  label: Ohio joint filing credit income threshold
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+description: Ohio provides a joint filing credit for this fraction  corresponds with the taxpayer's modified adjusted gross income, less exemptions for the taxable year, of the Tax less credits.
+metadata:
+  threshold_unit: currency-USD
+  rate_unit: /1
+  type: single_amount
+  label: Ohio joint filing credit income threshold
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml:6]:
```suggestion
  label: Ohio joint filing credit rate
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
+description: Ohio provides a joint filing credit for this fraction  corresponds with the taxpayer's modified adjusted gross income, less exemptions for the taxable year, of the Tax less credits.
+metadata:
+  threshold_unit: currency-USD
+  rate_unit: /1
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml:4]:
```suggestion
  amount_unit: /1
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+description: Ohio qualifies the joint filing credit for head and spouse each with Ohio adjusted gross income above this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml:1]:
```suggestion
description: Ohio qualifies filer for the joint filing credit with adjusted gross income above this amount. 
```

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+from policyengine_us.model_api import *
+
+
+class oh_joint_filing_credit(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Ohio Joint Filing Credit"
+    unit = USD
+    definition_period = YEAR
+    reference = "https://codes.ohio.gov/ohio-revised-code/section-5747.05"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py:10]: cant open link

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    defined_for = "oh_joint_filing_credit_eligible"
+
+    def formula(tax_unit, period, parameters):
+        tax = tax_unit("oh_income_tax_before_refundable_credits", period)
+        p = parameters(
+            period
+        ).gov.states.oh.tax.income.credits.joint_filing_credit
+        oh_agi = tax_unit("oh_agi", period)
+        exemption = tax_unit("oh_exemption", period)
+        magi_less_exepmtion = oh_agi - exemption
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py:20]: cap at 0

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+from policyengine_us.model_api import *
+
+
+class oh_joint_filing_credit_eligible(Variable):
+    value_type = bool
+    entity = TaxUnit
+    label = "Eligible for Ohio Joint Filing Credit"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:7]:
```suggestion
    label = "Eligible for the Ohio Joint Filing Credit"
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml [hunk 14]
```diff
@@ -0,0 +1,15 @@
+description: Ohio qualifies filer for the joint filing credit with adjusted gross income at or above this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml:1]:
```suggestion
description: Ohio qualifies filers for the joint filing credit with adjusted gross income at or above this amount. 
```

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    def formula(tax_unit, period, parameters):
+        filing_status = tax_unit("filing_status", period)
+        person = tax_unit.members
+        is_head = person("is_tax_unit_head", period)
+        is_spouse = person("is_tax_unit_spouse", period)
+        p = parameters(
+            period
+        ).gov.states.oh.tax.income.credits.joint_filing_credit
+        income = person("employment_income", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:20]: does it also include self-employment income?

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:20]: need to recreate 
oh_agi -> oh_alternative_agi
irs_gross_income -> oh_irs_gross_income
adjusted_gross_income -> oh_adjusted_gross_income

in the irs_gross_income we want to use a for loop for the income sources to exclude the income types as said on the form

File: policyengine_us/variables/gov/states/oh/tax/income/oh_exemption.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
+from policyengine_us.model_api import *
+
+
+class oh_exemption(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/oh_exemption.py:4]:
```suggestion
class oh_exemptions(Variable):
```

File: policyengine_us/variables/gov/states/oh/tax/income/oh_exemption.py [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+    # for reference
+    value_type = float
+    entity = TaxUnit
+    label = "Ohio exemption"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/oh_exemption.py:8]:
```suggestion
    label = "Ohio exemptions"
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
+from policyengine_us.model_api import *
+
+
+class oh_adjusted_gross_income(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py:4]:
```suggestion
class oh_alternative_adjusted_gross_income(Variable):
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "Ohio Alternative Adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py:7]:
```suggestion
    label = "Ohio Alternative Adjusted gross income for the joint filing credit"
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_alternative_agi.py [hunk 20]
```diff
@@ -0,0 +1,33 @@
+from policyengine_us.model_api import *
+
+
+class oh_alternative_agi(Variable):
+    value_type = float
+    entity = Person
+    label = "Ohio alternative adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_alternative_agi.py:7]:
```suggestion
    label = "Ohio alternative adjusted gross income for the joint filing credit"
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_irs_gross_income.py [hunk 21]
```diff
@@ -0,0 +1,31 @@
+from policyengine_us.model_api import *
+
+
+class oh_irs_gross_income(Variable):
+    value_type = float
+    entity = Person
+    label = "Ohio Gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_irs_gross_income.py:7]:
```suggestion
    label = "Ohio Gross income for the joint filing credit"
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py:1]: move these three files in the joint_filing_credit folder

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_adjusted_gross_income.py:1]: same with the tests

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml [hunk 22]
```diff
@@ -0,0 +1,15 @@
+description: Ohio qualifies filers for the joint filing credit with adjusted gross income at or above this amount.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/income_threshold.yaml:1]:
```suggestion
description: Ohio qualifies filers for the joint filing credit with alternative adjusted gross income at or above this amount. 
```

File: policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_alternative_agi.py [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://codes.ohio.gov/ohio-revised-code/section-5747.055",
+        "https://tax.ohio.gov/static/forms/ohio_individual/individual/2021/pit-it1040-booklet.pdf#page=20",
+    )
+    defined_for = StateCode.OH
+
+    adds = [
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/adjusted_gross_income/oh_alternative_agi.py:16]: can you make a list parameter - same for the subtracts part

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_agi.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_agi.py:1]: I dont think we need this as well as the alternative additions and subtractions - I thought we only need alternative gross income and federal agi? - Let me know if I am wrong

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: could you fix the merge conflicts? And re request my review

Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: tests don't pass

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_irs_gross_income.py [hunk 24]
```diff
@@ -0,0 +1,31 @@
+from policyengine_us.model_api import *
+
+
+class oh_irs_gross_income(Variable):
+    value_type = float
+    entity = Person
+    label = "Ohio gross income for the joint filing credit"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_irs_gross_income.py:7]:
```suggestion
    label = "Ohio gross income replicated for the Ohio joint filing credit"
```

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_agi.py [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+    )
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://codes.ohio.gov/ohio-revised-code/section-5747.055",
+        "https://tax.ohio.gov/static/forms/ohio_individual/individual/2021/pit-it1040-booklet.pdf#page=20",
+    )
+    defined_for = StateCode.OH
+
+    adds = ["oh_alternative_agi_additions"]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_agi.py:18]: does this not add adjusted_gross_income?

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_adjusted_gross_income.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_alternative_adjusted_gross_income.py:1]: make a folder called income for all the income types related to agi and gross income

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/oh/tax/income/oh_income_tax_before_non_refundable_credits.py [hunk 26]
```diff
@@ -1,7 +1,7 @@
 from policyengine_us.model_api import *
 
 
-class oh_income_tax_before_credits(Variable):
+class oh_income_tax_before_non_refundable_credits(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/oh_income_tax_before_non_refundable_credits.py:4]: need to change the tests and other variables accordingly

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: need a make format

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py [hunk 27]
```diff
@@ -0,0 +1,31 @@
+from policyengine_us.model_api import *
+
+
+class oh_joint_filing_credit_eligible(Variable):
+    value_type = bool
+    entity = TaxUnit
+    label = "Eligible for the Ohio Joint Filing Credit"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit_eligible.py:7]:
```suggestion
    label = "Eligible for the Ohio joint filing credit"
```

File: policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml [hunk 28]
```diff
@@ -0,0 +1,33 @@
+description: Ohio provides a joint filing credit for this fraction of qualified income.
+metadata:
+  threshold_unit: currency-USD
+  amount_unit: /1
+  type: single_amount
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/oh/tax/income/credits/joint_filing_credit/rate.yaml:5]:
```suggestion
  type: single_amount
  period: year
```

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

**Review by PavelMakarchuk — APPROVED** [review]

**Comment by PavelMakarchuk** [comment]
failed checks relate to labor supply responses.py

File: policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+    reference = "https://codes.ohio.gov/ohio-revised-code/section-5747.05"
+    defined_for = "oh_joint_filing_credit_eligible"
+
+    def formula(tax_unit, period, parameters):
+        tax_before_joint_filing_crdedit = tax_unit(
+            "oh_tax_before_joint_filing_credit", period
+        )
+        p = parameters(period).gov.states.oh.tax.income.credits.joint_filing
+        # Ohio use MAGI for the credit computation, which is Ohio AGI with
+        # the addition of the business income deduction, which is currently not included in the model,
```
[the remark this answers is not in the document]
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/oh/tax/income/credits/joint_filing_credit/oh_joint_filing_credit.py:19]: Filed #3527

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**